SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialTraceEventDataBuilder.cpp

```diff
@@ -0,0 +1,277 @@
+const char* FSpatialTraceEventName::SendCrossServerRPCEventName = GDK_EVENT_NAMESPACE "send_cross_server_rpc";
+const char* FSpatialTraceEventName::ReceiveCrossServerRPCEventName = GDK_EVENT_NAMESPACE "receive_cross_server_rpc";
+const char* FSpatialTraceEventName::MergePropertyUpdateEventName = GDK_EVENT_NAMESPACE "merge_property_update";
+const char* FSpatialTraceEventName::ReceiveCreateEntityEventName = GDK_EVENT_NAMESPACE "receive_create_entity";
+const char* FSpatialTraceEventName::ReceiveRemoveEntityEventName = GDK_EVENT_NAMESPACE "receive_remove_entity";
+const char* FSpatialTraceEventName::AuthorityChangeEventName = GDK_EVENT_NAMESPACE "authority_change";
+
+// ---- FStringCache ----
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::CombineStrings(const char* A, const char* B)
+{
+	if (NextIndex >= BufferSize)
+	{
+		return BufferSize - 1;
+	}
+
+	int32 InsertIndex = NextIndex;
+	int32 InitialRemainingSize = BufferSize - NextIndex;
+	FCStringAnsi::Strncpy(&Buffer[NextIndex], A, InitialRemainingSize);
+
+	int32 CharSize = sizeof(char);
+	int32 ALength = FCStringAnsi::Strlen(A);
+	int32 RemainingSize = InitialRemainingSize - ALength;
+	RemainingSize = FMath::Max(0, RemainingSize);
+
+	if (RemainingSize > CharSize)
+	{
+		FCStringAnsi::Strncpy(&Buffer[NextIndex + ALength], B, RemainingSize);
+	}
+
+	int32 BLength = FCStringAnsi::Strlen(B);
+	RemainingSize = RemainingSize - BLength - CharSize;
+	RemainingSize = FMath::Max(0, RemainingSize);
+
+	NextIndex += InitialRemainingSize - RemainingSize;
+	return InsertIndex;
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddString(const char* String)
+{
+	if (NextIndex >= BufferSize)
+	{
+		return BufferSize - 1;
+	}
+
+	int32 InsertIndex = NextIndex;
+	int32 InitialRemainingSize = BufferSize - NextIndex;
+	FCStringAnsi::Strncpy(&Buffer[NextIndex], String, InitialRemainingSize);
+
+	int32 CharSize = sizeof(char);
+	int32 StringLength = FCStringAnsi::Strlen(String);
+	int32 RemainingSize = InitialRemainingSize - StringLength - CharSize;
+	RemainingSize = FMath::Max(0, RemainingSize);
+
+	NextIndex += InitialRemainingSize - RemainingSize;
+	return InsertIndex;
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddFString(const FString& String)
+{
+	std::string ValueSrc = (const char*)TCHAR_TO_ANSI(*String);
+	return AddString(ValueSrc.c_str());
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddUInt32(uint32 Value)
+{
+	return AddInteger(Value, "%" PRIu32);
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddUInt64(uint64 Value)
+{
+	return AddInteger(Value, "%" PRIu64);
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddInt32(int32 Value)
+{
+	return AddInteger(Value, "%" PRId32);
+}
+
+int32 FSpatialTraceEventDataBuilder::FStringCache::AddInt64(int64 Value)
+{
+	return AddInteger(Value, "%" PRId64);
+}
+
+const char* FSpatialTraceEventDataBuilder::FStringCache::Get(int32 Handle) const
+{
+	if (Handle >= BufferSize || Handle < 0)
+	{
+		return nullptr;
```

Comment:
Is "" a better return type here?

Reply:
Unless worker is good with nullptr

Reply:
Worth noting that AddString always returns a valid handle

Reply:
True, I guess I was guarding the possibility someone types in whatever they want to the function. We can assume people are smart though and go for speeeed.

Reply:
Ah right, sorry didn't realize users could access this function - I suppose any function is available in the callback here, maybe we could make some private?